### PR TITLE
Use CPK to reduce the number of Tx necessary for market funding

### LINKED ISSUE #379: Cannot enter to markets with zero funding
Reproduction steps:

1. Create a new market
2. Remove all funding
3. Refresh I guess?

There's a "division by zero" error in the console so I guess the lack of funding is the problem.

### LINKED ISSUE #380: Use CPK to reduce the number of Tx necessary for market funding
Only adding funding, not removing it.

### PR DESCRIPTION
**Pull request by mariano-aguero**
Closes #380 
Closes #379 

#### Includes
- Add support to add funding with the CPK
- Add support to remove funding with the CPK

### CODE REVIEW

File: app/src/services/cpk.ts [hunk 1]
```diff
@@ -247,6 +253,98 @@ class CPKService {
...
+          operation: CPK.CALL,
+          to: collateral.address,
+          value: 0,
+          data: ERC20Service.encodeApproveUnlimited(account),
+        },
+        {
+          operation: CPK.CALL,
+          to: collateral.address,
+          value: 0,
+          data: ERC20Service.encodeTransferFrom(this.cpk.address, account, amount),
```
Comment (mariano-aguero) [inline app/src/services/cpk.ts]: Mmmm what do you think about this last transaction @fvictorio ?

File: app/src/util/tools.ts [hunk 2]
```diff
@@ -45,9 +45,15 @@ export const mulBN = (a: BigNumber, b: number, scale = 10000): BigNumber => {
...
-  const prices = holdings.map(holding => product.div(holding).div(denominator))
+  const denominator = holdings
+    .map(h => (h.toString() === '0' ? new Big(0) : product.div(h)))
+    .reduce((a, b) => a.add(b))
+
+  const prices = holdings.map(holding =>
+    product.toString() === '0' || denominator.toString() === '0'
+      ? new Big(0)
+      : product.div(holding).div(denominator),
+  )
```
Comment (mariano-aguero) [inline app/src/util/tools.ts]: I think this fix this problem https://github.com/protofire/gnosis-conditional-exchange/issues/379

File: app/src/services/cpk.ts [hunk 3]
```diff
@@ -325,6 +332,83 @@ class CPKService {
...
+          operation: CPK.CALL,
+          to: collateral.address,
+          value: 0,
+          data: ERC20Service.encodeTransferFrom(account, this.cpk.address, amount),
+        },
+        {
+          operation: CPK.CALL,
+          to: collateral.address,
+          value: 0,
+          data: ERC20Service.encodeApproveUnlimited(marketMaker.address),
```
Comment (fvictorio) [inline app/src/services/cpk.ts]: We should check if this transaction is necessary.

File: app/src/services/cpk.ts [hunk 4]
```diff
@@ -325,6 +332,83 @@ class CPKService {
...
+    marketMaker,
+  }: CPKFundingParams): Promise<TransactionReceipt> => {
+    try {
+      const transactions = [
+        {
+          operation: CPK.CALL,
+          to: collateral.address,
+          value: 0,
+          data: ERC20Service.encodeApproveUnlimited(this.cpk.address),
+        },
```
Comment (fvictorio) [inline app/src/services/cpk.ts]: Why are you doing this approval?

File: app/src/util/tools.ts (hunk 2 above)
Comment (fvictorio) [reply app/src/util/tools.ts]: This is probably fine for now, but the approach is unnecessarily involved. I would just add a check at the beginning of the function: if any holding is 0, then return all zeroes. (In practice, I think that either *all* holdings are zero, or all are non-zero, but anyway).

File: app/src/services/cpk.ts (hunk 3 above)
Comment (mariano-aguero) [reply app/src/services/cpk.ts]: I added a check for the allowance, and a push if is necessary

Comment (mariano-aguero) [reply app/src/services/cpk.ts]: Right here, this is a transaction with the approveUnlimited https://rinkeby.etherscan.io/tx/0x4ea69c9dd20dbfe268d1a139dbfc3ac4032bb5f55dcd74f7093d2c98c591fd6e#internal

Without the approveUnlimited
https://rinkeby.etherscan.io/tx/0x7b2529fa7f4e8a42f2ba816c0fe90b031511d46d90ef98488eedf00615f3746d#internal

File: app/src/services/cpk.ts (hunk 4 above)
Comment (mariano-aguero) [reply app/src/services/cpk.ts]: Mmmm is not necessary, I will remove it

File: app/src/util/tools.ts [hunk 5]
```diff
@@ -44,16 +44,16 @@ export const mulBN = (a: BigNumber, b: number, scale = 10000): BigNumber => {
  */
 export const calcPrice = (holdingsBN: BigNumber[]): number[] => {
   const holdings = holdingsBN.map(h => new Big(h.toString()))
+
+  const doesHoldingsAreZero = holdings.every(h => h.toString() === '0')
```
Comment (fvictorio) [inline app/src/util/tools.ts]: `hasZeroHoldings` or anything else with proper grammar.

Comment (mariano-aguero) [reply app/src/util/tools.ts]: I will fix this...

**Pull request merged**